eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs

```diff
@@ -1247,6 +1301,168 @@ submitFailingTxM tx mkExpectedFailures = do
+submitBlock_ = void . submitBlock
+
+-- | Submit a list of transactions as a block that's expected to succeed.
+-- The transactions are not modified to make them balance.
+submitBlock ::
+  ( HasCallStack
+  , ShelleyEraImp era
+  ) =>
+  [Tx TopTx era] -> ImpTestM era (Block BHeaderView era)
+submitBlock = expectRightDeepExpr . first fst <=< trySubmitBlock
+
+-- | Submit a list of transactions as a block that's expected to fail
+-- with the given predicate failures.
+-- The transactions are not modified to make them balance.
+submitFailingBlock ::
+  ( HasCallStack
+  , ShelleyEraImp era
+  ) =>
+  [Tx TopTx era] ->
+  NonEmpty (PredicateFailure (EraRule "BBODY" era)) ->
+  ImpTestM era ()
+submitFailingBlock txs = submitFailingBlockM txs . const . pure
+
+-- | Submit a list of transactions as a block that's expected to be rejected,
+-- and compute the expected predicate failures from the created block using the supplied action.
+-- The transactions are not modified to make them balance.
+submitFailingBlockM ::
+  ( HasCallStack
+  , ShelleyEraImp era
+  ) =>
+  [Tx TopTx era] ->
+  (Block BHeaderView era -> ImpTestM era (NonEmpty (PredicateFailure (EraRule "BBODY" era)))) ->
+  ImpTestM era ()
+submitFailingBlockM txs mkExpectedFailures = do
+  (predFailures, block) <- expectLeftDeepExpr =<< trySubmitBlock txs
+  expectedFailures <- mkExpectedFailures block
+  predFailures `shouldBeExpr` expectedFailures
+
+-- | Try to submit a list of transactions as a block,
+-- Return the block that was created using the transactions and any predicate
+-- failures that are produced.
+-- The transactions are not modified to make them balance.
+trySubmitBlock ::
+  forall era.
+  ShelleyEraImp era =>
+  [Tx TopTx era] ->
+  ImpTestM
+    era
+    ( Either
+        (NonEmpty (PredicateFailure (EraRule "BBODY" era)), Block BHeaderView era)
+        (Block BHeaderView era)
+    )
+trySubmitBlock (SSeq.fromList -> txs) = do
+  kh <- freshKeyHash
+  slotNo <- use impLastTickG
+  let blockHeader =
+        BHeaderView
+          { bhviewID = kh
+          , bhviewBSize = fromIntegral $ bBodySize (ProtVer (eraProtVerLow @era) 0) blockBody
+          , bhviewHSize = 0
+          , bhviewBHash = hashBlockBody blockBody
+          , bhviewSlot = slotNo
+          }
+      blockBody = mkBasicBlockBody @era & txSeqBlockBodyL .~ txs
+      block = Block {blockHeader, blockBody}
+
+  nes <- use impNESL
+  let ls = nes ^. nesEsL . esLStateL
+      pp = nes ^. nesEsL . curPParamsEpochStateL @era
+  res <-
+    tryRunImpRule @"BBODY"
+      (BbodyEnv pp (nes ^. chainAccountStateL))
+      (BbodyState ls (BlocksMade Map.empty))
+      block
+
+  case res of
+    Left predFailures -> do
+      -- Verify that produced predicate failures are ready for the node-to-client protocol
+      liftIO $ forM_ predFailures $ roundTripEraExpectation @era
+      pure $ Left (predFailures, block)
+    Right (BbodyState newState _, events) -> do
+      impNESL . nesEsL . esLStateL .= newState
```

Comment:
Where [`updateNewEpochState`](https://github.com/IntersectMBO/cardano-ledger/blob/512559b499059561c2b3ef1f1afe48bb384d2ea0/eras/shelley/impl/src/Cardano/Ledger/Shelley/API/Validation.hs#L217C1-L223)
```suggestion
    Right (bbodyState, events) -> do
      modifyNES (`updateNewEpochState` bbodyState)
```

Reply:
I haven't made this change yet since it now goes next to a change to the events, and I think the somewhat unintuitive things we're now doing with the state are clearer when expressed this way. However, I'm happy to add the change if you'd still prefer it.

Reply:
This isn't necessary now that I'm using `applyBlockEither` since that returns a `NewEpochState` that's already had `updateNewEpochState` applied to it